SSA/Core/Tactic.lean

```diff
@@ -15,115 +16,124 @@ open Ctxt (Var Valuation DerivedCtxt)
-        Ctxt.Valuation.snoc_last, Ctxt.map,
-        Ctxt.Valuation.snoc_eval, Ctxt.ofList, Ctxt.Valuation.snoc_toSnoc,
-        HVector.map, HVector.getN, HVector.get, HVector.toSingle, HVector.toPair, HVector.toTuple,
-        DialectDenote.denote, Expr.op_mk, Expr.args_mk,
-        DialectMorphism.mapOp, DialectMorphism.mapTy, List.map, Ctxt.snoc, List.map,
-        Function.comp, Valuation.ofPair, Valuation.ofHVector, Function.uncurry,
-        List.length_singleton, Fin.zero_eta, List.map_eq_map, List.map_cons, List.map_nil,
-        bind_assoc, pairBind,
-        /- `castPureToEff` -/
-        Com.letPure, Expr.denote_castPureToEff,
-        /- Unfold denotation -/
-        Com.denote_var, Com.denote_ret, Expr.denote_unfold, HVector.denote,
-        /- Effect massaging -/
-        EffectKind.toMonad_pure, EffectKind.toMonad_impure,
-        EffectKind.liftEffect_rfl,
-        Id.pure_eq, Id.bind_eq, id_eq,
-        /-
-        NOTE (Here Be Dragons 🐉): the parenthesis in `(HVector.denote_cons)` are significant!
-        `HVector.denote` has a typeclass assumption `TyDenote (Dialect.Ty d)`, where `d : Dialect`.
-        However, we tend to define `d` as an `abbrev`, so that our goal statement might have
-        `HVector.denote` where the concrete instance is `instTyDenote : TyDenote Ty`,
-          and `Ty` is the expression that `d` was defined with.
-
-        We've observed `simp [HVector.denote]` not working in such situations.
-          Even more surprising, `rw [HVector.denote]` *did* succeed in rewriting.
-        According to Zulip (https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/simp.20.5BX.5D.20fails.2C.20rw.20.5BX.5D.20works/near/358861409):
-        > simp [(X)] is a standard trick to fix simp [X] not working
-
-        By default, it seems that `simp` will synthesize typeclass arguments of a lemma, and then
-        use the *default* instance to determine whether a simp-lemma applies to the current goal.
-        Writing `simp [(X)]`, on the other hand, is equivalent to writing `simp [@X _ _ _]`
-          (for as many underscores as `X` takes arguments, implicit or explicit).
-        The parentheses seems to enable `simp` to unify typeclass arguments as well, and thus the
-          simp-lemma applies even for non-standard instances.
-
-        One caveat: `simp [(X)]` only works if `X` is a lemma, *not* if `X` is a definition to be
-        unfolded. Thus, we replace `HVector.denote` with its equation lemmas
-          `(HVector.denote_cons)` and `(HVector.denote_nil)` -/
-        (HVector.denote_cons), (HVector.denote_nil),
-        $ts,*]
-    ))
+  withLocalDeclsDND declInfo <| fun xs => do
+    let V := Valuation.mkOfElems Ty instTyDenote Γelems xs
+    let newType ← mkForallFVars xs (body.instantiate1 V)
+    let proof :=
+      let mp  :=
+        ←withLocalDeclD .anonymous e <| fun eProof => do
+          mkLambdaFVars #[eProof] <|← mkLambdaFVars xs <| mkApp eProof V
```

Comment:
```suggestion
          mkLambdaFVars (#[eProof] ++  xs) <| mkApp eProof V
```

I guess you opted against this because you didn't want to pay the price of reconstructing thearray because of `++`?

Reply:
That was indeed the reason, although looking at it now I can't imagine the performance difference is really a big deal. I'll go with your suggestion